plugin/src/lib.rs

```diff
@@ -496,6 +496,28 @@ impl<'a, 'cx> Folder for MutatorPlugin<'a, 'cx> {
                     attrs,
                 })
             }
+            Expr {
+                id,
+                node: ExprKind::WhileLet(pat, cond, block, opt_label),
+                span,
+                attrs,
+            } => {
+                let (n, current, sym, flag, mask) = self.add_mutations(
+                        cond.span,
+                        &["replacing whileLet Some() condition"],
+                    );
+               
+                let mut_block = quote_expr!(self.cx(), {
```

Comment:
You are almost there. Given your example:

```rust
while let Some(i) = optional {
        if i > 9 {
            println!("Greater than 9, quit!");
            optional = None;
        } else {
            println!("`i` is `{:?}`. Try again.", i);
            optional = Some(i + 1);
        }
    };
```

You will get:

```rust
while let Some(i) = if ::mutagen::now(1usize) { None } else {Some(        if i > 9 {
            println!("Greater than 9, quit!");
            optional = None;
        } else {
            println!("`i` is `{:?}`. Try again.", i);
            optional = Some(i + 1);
        }
)} {
        if i > 9 {
            println!("Greater than 9, quit!");
            optional = None;
        } else {
            println!("`i` is `{:?}`. Try again.", i);
            optional = Some(i + 1);
        }
    };
```

I think what you should change this to:

```rust
let mut_expr = quote_expr!(self.cx(), {
    if ::mutagen::now($n) { None } else { $cond }
});
```

Note that you should mutate the expression, which is on the right side. Note also that I've change the variable name on the snippet I propose and you should change it on line 516. With this changes, the example would expand to:


```rust
while let Some(i) = if ::mutagen::now(1usize) { None } else { optional )} {
        if i > 9 {
            println!("Greater than 9, quit!");
            optional = None;
        } else {
            println!("`i` is `{:?}`. Try again.", i);
            optional = Some(i + 1);
        }
    };
```

Reply:
And finally, there's a little last issue. Not all the `while let` expressions are used with `Option`. For example, consider this code:

```rust
fn test_result() {
    let mut optional = Ok(0);
    while let Ok(i) = optional {
        if i > 9 {
            println!("Greater than 9, quit!");
            optional = Err("end");
        } else {
            println!("`i` is `{:?}`. Try again.", i);
            optional = Ok(i + 1);
        }
    };
    assert!(optional == Err("end"));
}
```

Your code will make mutate those expressions and will lead to invalid code, as you will be comparing a `Result` with an `Option` (Look that you are changing the original expression to be `None`, which is of type `Option` and not `Result`).

What you can do is create a function that checks if  `pat` is of type `Some` (we don't have type information, but i think it's good enough, by now, to check if the arm is `Some`). If it does, you mutate the `while let`; otherwise, it should not mutate the expression and return the expression as-is.

Let me know if i'm not clear on some of the points or if you need some help or clarification.

Reply:
I think that something like this would work only for loops like: `while let Some(_) = $cond`, but not on `while let Ok(_) = $cond` or any other pattern.

To be honest, I'm not sure about how we should proceed with this, maybe @llogiq has some idea

Reply:
For the moment I think matching `Some` and `Ok` would be sufficient – and the mutation could use an internal AtomicBool to make the loop body apply once, so that we could have zero or one maximum iterations.

We can make the code generic by creating a trait in `mutagen`. The generated code could look like:

```
while let Some(_) = { static B: AtomicBool = ATOMIC_BOOL_INIT; ::mutagen::wl($expr, $n, &B, ...) } {
    ..
}
```

The `wl` function could then `fetch_or` the `AtomicBool` and generate `None` if the result is `true`. For `Result`s, it needs to `.map_err()` the original Result to `Result<T, ()>` (which is OK, since we throw away the error anyway) and generate `Result::Err(())` to stop the iteration instead.

Reply:
@llogiq I've to create new function plugin/lib.rs correct?